<a href="https://colab.research.google.com/github/VintaBytes/Ciencia-de-datos/blob/main/Ciencia%20De%20Datos%20Con%20Python/CuadernosColab/CienciaDeDatos_Cap%C3%ADtulo030_Comparar_categor%C3%ADas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Abrir en Colab"/></a>

# Capítulo 30 — Comparar categorías

## Comparar no es solo mirar qué valor es mayor

En los capítulos anteriores aprendimos a resumir datos por grupos usando `groupby()` y a construir tablas con varios indicadores.

Eso nos permite construir tablas con indicadores por categoría: cantidad de registros, totales, promedios y otros resúmenes.

Ahora vamos a concentrarnos en la comparación.

Comparar categorías no significa solamente buscar el número más alto. Primero tenemos que saber qué estamos comparando.

Por ejemplo, no es lo mismo preguntar:

- ¿qué día tiene más cuentas registradas?
- ¿qué día tiene mayor facturación total?
- ¿qué día tiene mayor promedio de cuenta?
- ¿qué día tiene mayor propina promedio?

Todas esas preguntas comparan días, pero no comparan lo mismo.

En este capítulo vamos a construir y leer comparaciones entre categorías, prestando atención a qué indicador usamos y qué conclusiones podemos sacar.

## Cargar el dataset de trabajo

Vamos a seguir usando el dataset `tips` de Seaborn.

Como ya conocemos sus columnas principales, podemos concentrarnos en comparar categorías a partir de distintos indicadores.

In [13]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("tips")

df.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


La tabla muestra las primeras filas del dataset.

Seguimos trabajando con las mismas variables de los capítulos anteriores. En este capítulo vamos a usar esas variables para comparar categorías.

Nos interesará especialmente observar cómo cambian algunos indicadores según columnas como `day` y `time`.

La idea no será solamente calcular valores, sino aprender a leer qué significa cada comparación.

## Construir una tabla para comparar días

Vamos a comenzar comparando los días registrados en el dataset.

La pregunta general será:

**¿Cómo se comportan las cuentas y las propinas según el día?**

Para responderla vamos a construir una tabla con varios indicadores por día:

- cantidad de cuentas;
- total facturado;
- promedio de cuenta;
- promedio de propina.

In [14]:
comparacion_dias = df.groupby(
    "day",
    observed=True
).agg(
    cantidad_cuentas=("total_bill", "count"),
    total_facturado=("total_bill", "sum"),
    promedio_cuenta=("total_bill", "mean"),
    promedio_propina=("tip", "mean")
)

comparacion_dias = comparacion_dias.round(2).reset_index()

comparacion_dias

,day,cantidad_cuentas,total_facturado,promedio_cuenta,promedio_propina
0,Thur,62,1096.33,17.68,2.77
1,Fri,19,325.88,17.15,2.73
2,Sat,87,1778.40,20.44,2.99
3,Sun,76,1627.16,21.41,3.26


La tabla permite comparar los días usando varios indicadores.

Podemos leer algunos resultados:

- `Sat` tiene la mayor cantidad de cuentas: **87**.
- `Sat` también tiene el mayor total facturado: **1778.40**.
- `Sun` tiene el promedio de cuenta más alto: **21.41**.
- `Sun` también tiene la propina promedio más alta: **3.26**.
- `Fri` tiene la menor cantidad de cuentas y el menor total facturado.

Esta tabla muestra una idea importante: el resultado de una comparación depende del indicador que usemos.

Si comparamos por facturación total, destaca `Sat`. Si comparamos por promedio de cuenta o promedio de propina, destaca `Sun`.

## Ordenar según distintos criterios

Para comparar categorías, muchas veces conviene ordenar la tabla según el indicador que queremos analizar.

Primero vamos a ordenar los días por `total_facturado`, de mayor a menor.

In [15]:
comparacion_dias.sort_values(
    by="total_facturado",
    ascending=False
)

,day,cantidad_cuentas,total_facturado,promedio_cuenta,promedio_propina
2,Sat,87,1778.40,20.44,2.99
3,Sun,76,1627.16,21.41,3.26
0,Thur,62,1096.33,17.68,2.77
1,Fri,19,325.88,17.15,2.73


Al ordenar por `total_facturado`, el primer lugar queda para `Sat`, con **1778.40**.

Luego aparecen:

- `Sun`, con **1627.16**;
- `Thur`, con **1096.33**;
- `Fri`, con **325.88**.

Este orden responde una pregunta concreta: **qué día acumuló más facturación total**.

Pero no necesariamente responde cuál fue el día con cuentas más altas en promedio. Para eso necesitamos ordenar por otro indicador.

## Comparar usando otro indicador

Ahora vamos a ordenar la misma tabla por `promedio_cuenta`.

Esto permite responder otra pregunta:

**¿Qué día tuvo el mayor promedio de cuenta?**

In [16]:
comparacion_dias.sort_values(
    by="promedio_cuenta",
    ascending=False
)

,day,cantidad_cuentas,total_facturado,promedio_cuenta,promedio_propina
3,Sun,76,1627.16,21.41,3.26
2,Sat,87,1778.40,20.44,2.99
0,Thur,62,1096.33,17.68,2.77
1,Fri,19,325.88,17.15,2.73


Al ordenar por `promedio_cuenta`, el primer lugar cambia.

Ahora aparece `Sun`, con un promedio de cuenta de **21.41**. Luego aparecen:

- `Sat`, con **20.44**;
- `Thur`, con **17.68**;
- `Fri`, con **17.15**.

Esto muestra que `Sat` acumuló más facturación total, pero `Sun` tuvo cuentas más altas en promedio.

La diferencia es importante:

- `total_facturado` depende de cuánto se consume y de cuántas cuentas hay.
- `promedio_cuenta` resume el valor medio de cada cuenta.

Por eso, antes de decir que una categoría “rinde más” o “es mejor”, tenemos que aclarar qué indicador estamos usando.

## Comparar propinas promedio

También podemos ordenar los días según la propina promedio.

Esta comparación responde otra pregunta:

**¿Qué día tuvo la mayor propina promedio?**

In [17]:
comparacion_dias.sort_values(
    by="promedio_propina",
    ascending=False
)

,day,cantidad_cuentas,total_facturado,promedio_cuenta,promedio_propina
3,Sun,76,1627.16,21.41,3.26
2,Sat,87,1778.40,20.44,2.99
0,Thur,62,1096.33,17.68,2.77
1,Fri,19,325.88,17.15,2.73


Al ordenar por `promedio_propina`, el orden queda igual que al ordenar por `promedio_cuenta`.

`Sun` aparece en primer lugar, con una propina promedio de **3.26**. Luego aparecen:

- `Sat`, con **2.99**;
- `Thur`, con **2.77**;
- `Fri`, con **2.73**.

En este caso, el día con mayor promedio de cuenta también tiene la mayor propina promedio.

Esto puede sugerir una relación entre cuentas más altas y propinas más altas, pero todavía no alcanza para afirmarlo con seguridad. Por ahora solo estamos comparando promedios por día.

## Escribir una comparación con cuidado

Una tabla puede mostrar diferencias entre categorías, pero la interpretación escrita debe ser prudente.

No es lo mismo decir:

**“El domingo es el mejor día.”**

que decir:

**“En este dataset, el domingo tiene el mayor promedio de cuenta y la mayor propina promedio.”**

La segunda frase es más precisa, porque indica:

- que estamos hablando de este dataset;
- qué categoría estamos observando;
- qué indicadores usamos para comparar;
- qué resultado concreto encontramos.

Cuando comunicamos un análisis, conviene evitar conclusiones demasiado generales si los datos solo permiten una comparación específica.

## Comparar horarios

Ahora vamos a comparar otra categoría: el horario.

La pregunta será:

**¿Cómo cambian las cuentas y las propinas entre almuerzo y cena?**

Vamos a construir una tabla similar a la anterior, pero agrupando por `time`.

In [18]:
comparacion_horarios = df.groupby(
    "time",
    observed=True
).agg(
    cantidad_cuentas=("total_bill", "count"),
    total_facturado=("total_bill", "sum"),
    promedio_cuenta=("total_bill", "mean"),
    promedio_propina=("tip", "mean")
)

comparacion_horarios = comparacion_horarios.round(2).reset_index()

comparacion_horarios

,time,cantidad_cuentas,total_facturado,promedio_cuenta,promedio_propina
0,Lunch,68,1167.47,17.17,2.73
1,Dinner,176,3660.30,20.80,3.10


La tabla compara los dos horarios registrados en el dataset.

Podemos observar que:

- `Dinner` tiene **176 cuentas**, mientras que `Lunch` tiene **68**.
- `Dinner` acumula una facturación total de **3660.30**, frente a **1167.47** en `Lunch`.
- El promedio de cuenta también es mayor en `Dinner`: **20.80** contra **17.17**.
- La propina promedio es mayor en `Dinner`: **3.10** contra **2.73**.

En este caso, `Dinner` aparece por encima de `Lunch` en todos los indicadores calculados.

De todos modos, la interpretación debe mencionar qué se está comparando. Podemos decir que, en este dataset, la cena concentra más registros, más facturación total y mayores promedios de cuenta y propina.

## De la tabla a una interpretación

Una tabla de comparación no debe leerse solo buscando el número más grande.

Conviene preguntarse:

- ¿qué indicador estoy comparando?
- ¿los grupos tienen cantidades de registros parecidas?
- ¿el resultado habla de totales o de promedios?
- ¿la conclusión está limitada al dataset analizado?

Por ejemplo, a partir de las tablas anteriores podemos escribir:

**En este dataset, el sábado concentra la mayor facturación total por día, mientras que el domingo presenta el mayor promedio de cuenta y la mayor propina promedio. Al comparar horarios, la cena supera al almuerzo tanto en cantidad de registros como en facturación total, promedio de cuenta y promedio de propina.**

Esta interpretación es más clara y más prudente que decir simplemente “sábado es mejor” o “la cena es mejor”, porque indica qué se comparó y con qué indicadores.

## Cierre del capítulo

En este capítulo usamos tablas agrupadas para comparar categorías.

Vimos que comparar no significa solamente buscar el valor más alto. Primero hay que definir qué indicador estamos usando: cantidad de registros, total facturado, promedio de cuenta o promedio de propina.

También vimos que distintos indicadores pueden llevar a lecturas distintas. Un grupo puede tener mayor facturación total porque tiene más registros, mientras que otro puede tener un promedio más alto por cuenta.

Por eso, al comparar categorías, conviene leer varios indicadores juntos y escribir conclusiones prudentes, limitadas al dataset analizado.

En los próximos capítulos vamos a seguir trabajando con comparaciones, rankings y valores extremos, para detectar categorías destacadas y ordenar resultados de manera más precisa.